OpenRA.Game/Traits/Player/FrozenActorLayer.cs

```diff
@@ -226,24 +224,10 @@ public FrozenActorLayer(Actor self, FrozenActorLayerInfo info)
 			owner = self.Owner;
 			frozenActorsById = new Dictionary<uint, FrozenActor>();
 
-			// PERF: Partition the map into a series of coarse-grained bins and track changes in the shroud against
-			// bin - marking that bin dirty if it changes. This is fairly cheap to track and allows us to perform the
-			// expensive visibility update for frozen actors in these regions.
 			partitionedFrozenActorIds = new SpatiallyPartitioned<uint>(
 				world.Map.MapSize.X, world.Map.MapSize.Y, binSize);
 
-			cols = world.Map.MapSize.X / binSize + 1;
-			rows = world.Map.MapSize.Y / binSize + 1;
-			dirtyBins = new bool[cols * rows];
-			self.Trait<Shroud>().CellsChanged += cells =>
-			{
-				foreach (var cell in cells)
-				{
-					var x = cell.U / binSize;
-					var y = cell.V / binSize;
-					dirtyBins[y * cols + x] = true;
-				}
-			};
+			self.Trait<Shroud>().OnShroudChanged += uv => dirtyFrozenActorIds.UnionWith(partitionedFrozenActorIds.At(new int2(uv.U, uv.V)));
```

Comment:
Previously we were checking a region via `InBox` due to the coarse binning scheme in `FrozenActorLayer` that went away, so this would be an improvement anyway - certainly no worse at least.